Refactor k-NN filter search

kolchfa-aws opened the pull request:
Fixes #3221 

### Checklist
- [x] By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and subject to the [Developers Certificate of Origin](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

martin-gaievski commented on _search-plugins/knn/filter-search-knn.md:
```diff
@@ -8,119 +8,84 @@ has_math: true
+- [Boolean filter](#boolean-filter-with-ann-search): This approach runs an [Approximate nearest neighbor (ANN)]({{site.url}}{{site.baseurl}}/search-plugins/knn/approximate-knn) search, and then applies the filter to the results. Because it uses post-filtering, a query with a Boolean filter may return significantly less than `k` results for a restrictive filter.
 
-The OpenSearch k-NN plugin version 2.2 introduced support for the Lucene engine in order to process k-NN searches. The Lucene engine provides a search that is based on the HNSW algorithm in order to represent a multi-layered graph. The OpenSearch k-NN plugin version 2.4 can incorporate filters for searches based on Lucene 9.4.
+- [Lucene k-NN filter](#using-a-lucene-k-nn-filter): This approach ensures that `k` results are returned because filtering is applied during the k-NN search as opposed to after the k-NN search, like in post-filtering. The drawback is that you can only use this method with the Hierarchical Navigable Small World (HNSW) algorithm implemented by a Lucene search engine.
```
Please mention that lucene filtering is available from 2.4.

martin-gaievski commented on _search-plugins/knn/filter-search-knn.md:
```diff
@@ -8,119 +8,84 @@ has_math: true
+## Filtered search optimization
 
-Lucene also provides the capability to operate its `KnnVectorQuery` across a subset of documents. To learn more about this capability, see the [Apache Lucene Documentation](https://issues.apache.org/jira/browse/LUCENE-10382).
+Overall, Lucene k-NN filters are more efficient compared to other filtering methods, both in terms of performance and relevancy of search results. 
```
This statement isn't correct for all use cases, and later we provide table where each of 3 filtering types paired with the dataset and filter combination.

martin-gaievski commented on _search-plugins/knn/filter-search-knn.md:
```diff
@@ -8,119 +8,84 @@ has_math: true
+Overall, Lucene k-NN filters are more efficient compared to other filtering methods, both in terms of performance and relevancy of search results. 
 
-To learn more about all available k-NN search approaches, including approximate k-NN, exact k-NN with script score, and pre-filtering with painless extensions, see [k-NN]({{site.url}}{{site.baseurl}}/search-plugins/knn/index/).
+Depending on your dataset and use case, you might be more interested in maximizing recall or minimizing latency. The following table provides guidance on various k-NN search configurations and the filtering methods to use in order to optimize for better recall or lower latency. The first three columns of the table provide several example k-NN search configurations. A search configuration consists ...
```
I think we need to state "higher recall" if it follows by "lower latency".

martin-gaievski commented on _search-plugins/knn/filter-search-knn.md:
```diff
@@ -8,119 +8,84 @@ has_math: true
-
-Number of Vectors | Filter Restrictive Percentage | k | Recall | Latency
--- | -- | -- | -- | --
+Number of documents in an index | Percentage of documents the filter returns | k | Filtering method to use for best recall | Filtering method to use for lowest latency
```
same as above, not sure we need to say "best recall", probably "higher recall" is more accurate

kolchfa-aws commented:
@martin-gaievski I have added the scoring script example (changed the query_value to use a double array because integers raised an exception) and addressed the comments. Please re-review when you get a chance and approve the PR when everything looks good. Thanks!

martin-gaievski reviewed (approved):
Looks good to me, thank you

cwillum commented on _search-plugins/knn/filter-search-knn.md:
```diff
@@ -1,126 +1,140 @@
-
-Number of Vectors | Filter Restrictive Percentage | k | Recall | Latency
--- | -- | -- | -- | --
+Number of documents in an index | Percentage of documents the filter returns | k | Filtering method to use for higher recall | Filtering method to use for lower latency
```
Jeff has got everyone going on this program: add pipes before and after all table elements (in case we ever switch to a host that can only interpret a markdown table with this format. Not sure if you think that's worthwhile.

cwillum reviewed (approved):
Looks great.

natebower commented on _search-plugins/knn/filter-search-knn.md:
```diff
@@ -1,126 +1,140 @@
+- [Scoring script filter](#scoring-script-filter): This approach involves pre-filtering a document set and then running an exact k-NN search on the filtered subset. It does not scale for large filtered subsets.
 
-## How does a k-NN filter work?
+- [Boolean filter](#boolean-filter-with-ann-search): This approach runs an [Approximate nearest neighbor (ANN)]({{site.url}}{{site.baseurl}}/search-plugins/knn/approximate-knn) search and then applies a filter to the results. Because of post-filtering, it may return significantly less than `k` results for a restrictive filter.
```
```suggestion
- [Boolean filter](#boolean-filter-with-ann-search): This approach runs an [approximate nearest neighbor (ANN)]({{site.url}}{{site.baseurl}}/search-plugins/knn/approximate-knn) search and then applies a filter to the results. Because of post-filtering, it may return significantly fewer than `k` results for a restrictive filter.
```

natebower commented on _search-plugins/knn/filter-search-knn.md:
```diff
@@ -1,126 +1,140 @@
+- [Boolean filter](#boolean-filter-with-ann-search): This approach runs an [Approximate nearest neighbor (ANN)]({{site.url}}{{site.baseurl}}/search-plugins/knn/approximate-knn) search and then applies a filter to the results. Because of post-filtering, it may return significantly less than `k` results for a restrictive filter.
 
-The OpenSearch k-NN plugin version 2.2 introduced support for the Lucene engine in order to process k-NN searches. The Lucene engine provides a search that is based on the HNSW algorithm in order to represent a multi-layered graph. The OpenSearch k-NN plugin version 2.4 can incorporate filters for searches based on Lucene 9.4.
+- [Lucene k-NN filter](#using-a-lucene-k-nn-filter): This approach applies filtering _during_ the k-NN search, as opposed to before or after the k-NN search, which ensures that `k` results are returned. You can only use this method with the Hierarchical Navigable Small World (HNSW) algorithm implemented by the Lucene search engine in k-NN plugin versions 2.4 and above.
```
```suggestion
- [Lucene k-NN filter](#using-a-lucene-k-nn-filter): This approach applies filtering _during_ the k-NN search, as opposed to before or after the k-NN search, which ensures that `k` results are returned. You can only use this method with the Hierarchical Navigable Small World (HNSW) algorithm implemented by the Lucene search engine in k-NN plugin versions 2.4 and later.
```

natebower commented on _search-plugins/knn/filter-search-knn.md:
```diff
@@ -1,126 +1,140 @@
+## Filtered search optimization
 
-Lucene also provides the capability to operate its `KnnVectorQuery` across a subset of documents. To learn more about this capability, see the [Apache Lucene Documentation](https://issues.apache.org/jira/browse/LUCENE-10382).
+Depending on your dataset and use case, you might be more interested in maximizing recall or minimizing latency. The following table provides guidance on various k-NN search configurations and the filtering methods to use in order to optimize for higher recall or lower latency. The first three columns of the table provide several example k-NN search configurations. A search configuration consists ...
```
```suggestion
Depending on your dataset and use case, you might be more interested in maximizing recall or minimizing latency. The following table provides guidance on various k-NN search configurations and the filtering methods used to optimize for higher recall or lower latency. The first three columns of the table provide several example k-NN search configurations. A search configuration consists of:
```

natebower commented on _search-plugins/knn/filter-search-knn.md:
```diff
@@ -1,126 +1,140 @@
 
-To learn more about all available k-NN search approaches, including approximate k-NN, exact k-NN with script score, and pre-filtering with painless extensions, see [k-NN]({{site.url}}{{site.baseurl}}/search-plugins/knn/index/).
+- The number of documents in an index, where one OpenSearch document corresponds to one k-NN vector.
+- The percentage of documents left in the results after filtering. This value depends on the restrictiveness of the filter that you are providing in the query. The most restrictive filter in the table returns 2.5% of documents in the index, while the least restrictive filter returns 80% of documents.
```
```suggestion
- The percentage of documents left in the results after filtering. This value depends on the restrictiveness of the filter that you provide in the query. The most restrictive filter in the table returns 2.5% of documents in the index, while the least restrictive filter returns 80% of documents.
```

natebower commented on _search-plugins/knn/filter-search-knn.md:
```diff
@@ -1,126 +1,140 @@
+| 10M | 38 | 100 | Lucene filter | Boolean filter |
+| 10M | 80 | 100 | Scoring script | Lucene filter |
+| 1M | 2.5 | 100 | Lucene filter | Scoring script |
+| 1M | 38 | 100 | Lucene filter | Lucene filter/Scoring script |
```
```suggestion
| 1M | 38 | 100 | Lucene filter | Lucene filter/scoring script |
```

natebower commented on _search-plugins/knn/filter-search-knn.md:
```diff
@@ -1,126 +1,140 @@
+## Scoring script filter
 
-![How the algorithm evaluates a doc set]({{site.url}}{{site.baseurl}}/images/hsnw-algorithm.png)
+A scoring script filter first filters the documents and then uses a brute-force exact k-NN search on the results. For example, the following query searches for hotels with a rating between 8 and 10, inclusive, that provide parking, and then performs a k-NN search to return the three hotels that are closest to the specified `location`:
```
```suggestion
A scoring script filter first filters the documents and then uses a brute-force exact k-NN search on the results. For example, the following query searches for hotels with a rating between 8 and 10, inclusive, that provide parking and then performs a k-NN search to return the 3 hotels that are closest to the specified `location`:
```

natebower commented on _search-plugins/knn/filter-search-knn.md:
```diff
@@ -1,126 +1,140 @@
-#### Example response
 
-The Boolean query filter returns the following results in the response:
+The response contains documents with the matching hotels:
```
```suggestion
The response includes documents containing the matching hotels:
```

natebower commented on _search-plugins/knn/filter-search-knn.md:
```diff
@@ -183,167 +197,73 @@ The Boolean query filter returns the following results in the response:
+## Lucene k-NN filter implementation
 
-A very restrictive filter returns the lowest number of documents in your dataset. For example, the following filter criteria specifies hotels with feedback ratings less than or equal to 3. This 2.5% filter only returns 1 document:
+The OpenSearch k-NN plugin version 2.2 introduced support for running k-NN searches with the Lucene engine using HNSW graphs. Starting with version 2.4, which is based on Lucene version 9.4, you can use Lucene filters for k-NN searches.
```
```suggestion
k-NN plugin version 2.2 introduced support for running k-NN searches with the Lucene engine using HNSW graphs. Starting with version 2.4, which is based on Lucene version 9.4, you can use Lucene filters for k-NN searches.
```

natebower commented on _search-plugins/knn/filter-search-knn.md:
```diff
@@ -183,167 +197,73 @@ The Boolean query filter returns the following results in the response:
+- k: The maximum number of vectors to return in the response.
 
-A somewhat restrictive filter returns 38% of the documents in the data set that you search. For example, the following filter criteria specifies hotels with parking and feedback ratings less than or equal to 8 and returns 5 documents:
+The following flow chart summarizes the Lucene algorithm.
```
```suggestion
The following flow chart outlines the Lucene algorithm.
```

natebower commented on _search-plugins/knn/filter-search-knn.md:
```diff
@@ -183,167 +197,73 @@ The Boolean query filter returns the following results in the response:
+![Lucene algorithm for filtering]({{site.url}}{{site.baseurl}}/images/lucene-algorithm.png)
 
-### Use case 3: Not very restrictive 80% filter
+For more information about the Lucene filtering implementation and the underlying `KnnVectorQuery`, see the [Apache Lucene Documentation](https://issues.apache.org/jira/browse/LUCENE-10382).
```
```suggestion
For more information about the Lucene filtering implementation and the underlying `KnnVectorQuery`, see the [Apache Lucene documentation](https://issues.apache.org/jira/browse/LUCENE-10382).
```

natebower reviewed (commented):
@kolchfa-aws Please see my comments and changes and let me know if you have any questions. Thanks!
